src/funnel.ts

```diff
@@ -0,0 +1,280 @@
+ * to ensure that everything _else_ is well-typed.
+ *
+ * Notice that this function constructs a funnel **object**, and does **not**
+ * execute anything when called. The returned object should be used to execute
+ * the funnel via the its `call` method.
+ *
+ * Debouncing: use `burstCoolDownMs` and any `invokedAt`.
+ * Throttling: use `delayMs` and `invokedAt: "start"` or `"both"`.
+ * Batching: See the reference implementation in [`funnel.reference-batch.test.ts`](https://github.com/remeda/remeda/blob/main/src/funnel.reference-batch.test.ts).
+ *
+ * @param callback - The main function that would be invoked occasionally based
+ * on `options`. The function would take the latest result of the `reducer`; if
+ * no calls where made since the last time it was invoked it will not be
+ * invoked. If a return value is needed, it should be passed via a reference or
+ * via closure to the outer scope of the funnel.
+ * @param options - An object that defines when `execute` should be
+ * invoked, relative to the calls of `call`. A timer that isn't defined will
+ * **not** be enabled, 0 is not used a default fallback.
+ * @param options.reducer - Combines the arguments passed to `call` with the
+ * value computed on the previous call (or `undefined` on the first time). The
+ * goal of the function is to extract and summarize the data needed for
+ * `callback`. It should be fast and simple as it is called often and should
+ * defer heavy operations to the `execute` function. If the final value
+ * is `undefined`, `callback` will not be called.
+ * @param options.invokedAt - At what "edges" of the funnel's activity
+ * window should `execute` be invoked. `start` means The function will be
+ * invoked immediately (withing the **same** execution frame!), and any
+ * subsequent calls would be ignored until the funnel is idle again. During
+ * this period `reducer` will also not be called. `end` The function will
+ * **not** be invoked initially but the timer will be started. Any calls during
+ * this time would be passed to the reducer, and when the timers are done, the
+ * reduced result would trigger an invocation. When `both` is used The function
+ * will be invoked immediately, and then the funnel would behave as if it was
+ * in the 'end' state. @default 'end'.
+ * @param options.burstCoolDownMs - The burst timer prevents subsequent
+ * calls in short succession to cause excessive invocations (aka "debounce").
+ * This duration represents the **minimum** amount of time that needs to pass
+ * between calls (the "quiet" part) in order for the subsequent call to **not**
+ * be considered part of the burst. In other words, as long as calls are faster
+ * than this, they are considered part of the burst.
+ * @param options.maxBurstDurationMs - Bursts are extended every time a
+ * call is made within the burst period. This means that the burst period could
+ * be extended indefinitely. To prevent such cases, a maximum burst duration
+ * could be defined.
```

Comment:
**note:** i want to bikeshed how to name this and describe this... at minimum, if we're sticking with `burstCooldownMs` and `maxBurstDurationMs`, we should actually give an explicit definition of a burst

Reply:
I don't mind considering other names.

I've said in the past that I don't want remeda to just be a shallow clone of Lodash, and we should consider improving in places where Lodash didn't do a good job; Lodash uses 'wait' and 'maxWait' and I felt that those don't represent the semantics well when we add a third time param (delay) to the mix. I didn't want to call these 'debounceTime' and 'throttleTime' because that requires someone to know what debounce and throttle is. I tried to find names that make the most sense when Remeda is the only library you know.